File: src/providers/openrouter.ts
import type { Provider } from "../core/types.js";

export const openrouter: Provider = {
  name: "openrouter",
  displayName: "OpenRouter",
  keyEnvVar: "OPENROUTER_API_KEY",
  detect: (key) => /^sk-or-(v\d-)?[A-Za-z0-9]{32,}$/.test(key),
  validate: async (key, { http, signal }) => {
    const res = await http("https://openrouter.ai/api/v1/auth/key", {
      headers: { Authorization: `Bearer ${key}` },
      signal,
    });
    if (!res.ok) return { status: res.status, ok: false };
    const body = (await res.json().catch(() => ({}))) as {
      data?: { label?: string; usage?: number; limit?: number | null };
    };
    return {
      status: res.status,
      ok: true,
      metadata: body.data
        ? { label: body.data.label, usage: body.data.usage, limit: body.data.limit }
        : undefined,
    };
  },
};

File: python/src/llm_key_validator/providers/openrouter.py
from __future__ import annotations

import re

from ..core.types import Provider, ProviderRawResult, ValidationContext

_DETECT = re.compile(r"^sk-or-(v\d-)?[A-Za-z0-9]{32,}$")


def _validate(key: str, ctx: ValidationContext) -> ProviderRawResult:
    res = ctx.http(
        "https://openrouter.ai/api/v1/auth/key",
        headers={"Authorization": f"Bearer {key}"},
    )
    if not res.ok:
        return ProviderRawResult(status=res.status, ok=False)
    body = res.json() or {}
    data = body.get("data") if isinstance(body, dict) else None
    metadata = None
    if isinstance(data, dict):
        metadata = {
            "label": data.get("label"),
            "usage": data.get("usage"),
            "limit": data.get("limit"),
        }
    return ProviderRawResult(status=res.status, ok=True, metadata=metadata)


openrouter = Provider(
    name="openrouter",
    display_name="OpenRouter",
    key_env_var="OPENROUTER_API_KEY",
    detect=lambda k: bool(_DETECT.match(k)),
    validate=_validate,
)
